dlt/destinations/synapse/sql_client.py

```diff
@@ -0,0 +1,342 @@
+        table_name = sql.split(" INTO ")[1].split("(")[0].strip()
+        column_names_str = sql.split("(")[1].split(")")[0].strip()
+
+        # Extracting row data
+        row_data_str = sql.split("VALUES")[1].strip()
+        # Using regex to split rows accurately
+        rows = [tuple(re.split(r"\s*,\s*(?![^()]*\))", row)) for row in re.findall(r"\((.*?)\)", row_data_str)]
+
+        # Ensure each item in rows is properly formatted
+        formatted_rows = []
+        for row in rows:
+            formatted_row = []
+            for item in row:
+                # Remove extra single quotes if item is not a JSON string
+                if item.startswith("'") and item.endswith("'") and not item.lstrip("'").startswith("{"):
+                    formatted_row.append(item.strip("'"))
+                else:
+                    formatted_row.append(item)
+            formatted_rows.append(tuple(formatted_row))
+
+        # Building SELECT statements with parameter markers
+        select_statements = [f"SELECT {', '.join(['?' for _ in row])}" for row in formatted_rows]
+
+        # Combining SELECT statements with UNION ALL
+        all_select_statements = " UNION ALL ".join(select_statements)
+
+        # Building the final SQL query
+        new_sql = f'INSERT INTO {table_name}({column_names_str}) {all_select_statements}'
+
+        # Extracting parameter values
+        param_values = [item for row in formatted_rows for item in row]
+
+        #logging.info(f'Generated SQL: {new_sql}')
+        #logging.info(f'Param values: {param_values}')
+
+        return new_sql, param_values
+
+    def open_connection(self) -> pyodbc.Connection:
+        try:
+            # Establish a connection
+            conn_str = (
+                f"DRIVER={{{self.credentials.odbc_driver}}};"
+                f"SERVER={self.credentials.host};"
+                f"DATABASE={self.credentials.database};"
+                f"UID={self.credentials.user};"
+                f"PWD={self.credentials.password};"
+            )
+            self._conn = pyodbc.connect(conn_str)
+
+            # Add the converter for datetimeoffset
+            self._conn.add_output_converter(-155, handle_datetimeoffset)
+
+            # Noting that autocommit is being set to True
+            self._conn.autocommit = True
+
+            return self._conn
+
+        except pyodbc.Error as e:
+            raise  # re-raise the error without logging it
+
+
+    @raise_open_connection_error
+    def close_connection(self) -> None:
+        try:
+            if self._conn:
+                if self._transaction_in_progress:
+                    self._conn.commit()
+                self._conn.close()
+        except pyodbc.Error as e:
+            # Log detailed error information here
+            logger.error(f"Failed to close connection due to pyodbc.Error: {str(e)}")
+
+
+    @contextmanager
+    def begin_transaction(self) -> Iterator[DBTransaction]:
```

Comment:
what exactly happens when we use the same tx code we have in the mssql implementation? why can't we use that?

Reply:
Shifted back to mssql implementation.  Current approach helped with error handling; however, we aren't facing transaction errors anymore and can switch back.